### ISSUE TITLE
Noisy Imagenette/Woof Proposal

### ISSUE THREAD

**Issue by tmabraham** [fastai/imagenette#40]
# Noisy Imagenette/Woof

## Introduction
Most of the time, dataset labels are actually quite noisy as the humans generating those labels are error-prone. This is especially the case for labels generated through crowdsourcing. Recently, there has been significant research in dealing with noisy labels in datasets and training deep learning models that are robust to noise (ex: [here](https://openreview.net/forum?id=ZPa2SyGcbwh), [here](https://arxiv.org/abs/2007.00151), [here](https://arxiv.org/abs/1906.03361), [here](https://arxiv.org/abs/1906.03361), [here](https://arxiv.org/abs/2010.01412), and [here](https://openreview.net/forum?id=D1E1h-K3jso)). It would be great to be able to implement some of these techniques in fastai (some work has already been done on this front, ex: [here](https://www.kaggle.com/keremt/progressive-label-correction-paper-implementation)) and test them on a benchmarking dataset. 

## Proposal
I propose to add to this dataset/repository a corrupted version of Imagenette and ImageWoof where the training labels are switched randomly at varying probabilities to simulate datasets with different levels of noise. This dataset is currently available [here](https://github.com/tmabraham/noisy_imagenette). The images themselves are the same as Imagenette, but the labels are instead provided in a [CSV file](https://github.com/tmabraham/noisy_imagenette/blob/main/noisy_imagenette.csv). There are 4 noise levels: 1%, 5%, 25%, 50% noise. The generation of these labels are provided [here](https://github.com/tmabraham/noisy_imagenette/blob/main/generate_labels.ipynb). A [baseline](https://github.com/tmabraham/noisy_imagenette/blob/main/baseline/baseline-01-18-2021.md) based on the original Imagenette baseline is [provided as well](https://github.com/tmabraham/noisy_imagenette/blob/main/train_noisy_imagenette.py). 

Note that in this `imagenette` repository there are currently 24 leaderboards (12 for Imagenette and 12 for ImageWoof). Since there are 4 noise levels, there is a total of 96 possible leaderboards. I have run my baseline for all 96 possibilities (done automatically on 4 TITAN RTXs with [this bash script](https://github.com/tmabraham/noisy_imagenette/blob/main/baseline/run_leaderboard.sh)) and provided an extended leaderboard over [here](https://github.com/tmabraham/noisy_imagenette/blob/main/extended_lb.csv). I have also selected 16 leaderboards that I have kept on the README. It is up to the maintainers (Jeremy and Hamel) to decide which leaderboards they want to keep.

## What needs to be done
These are the things I believe needs to be done for this dataset to be added to Imagenette:
- [ ] Add URLs for Noisy Imagenette/Woof (same as Imagenette but with the extra CSV file)
- [ ] Select the leaderboards to include 
- [ ] Add the leaderboards to the repository README
- [ ] Potentially add the label generation, benchmarks, extended LB in another folder on this repository? Alternatively, I could keep it hosted in my repository.

@jph00 and @hamelsmu Please review my [repository](https://github.com/tmabraham/noisy_imagenette) and let me know if you think there are any additional tasks that I would need to do.

**Comment by hamelsmu** [comment]
As far as URLs are concerned I think @jph00 can upload this to the appropriate location https://github.com/tmabraham/noisy_imagenette/blob/main/noisy_imagenette.csv

Tanishq, do you want to start with a proposal on which leaderboards you think might be best to keep or highlight?  You would probably know better than me at this point

**Comment by jph00** [comment]
> As far as URLs are concerned I think @jph00 <https://github.com/jph00> can upload this to the appropriate location https://github.com/tmabraham/noisy_imagenette/blob/main/noisy_imagenette.csv


I think it's more than just uploading the CSV - I think we should update the tgz to include the csv, and update the example training script to use the csv too. That way people can easily change to different labels by just changing the column they select in the csv.

I can provide the upload details when you're ready to upload.

**Comment by tmabraham** [comment]
@hamelsmu The [README](https://github.com/tmabraham/noisy_imagenette/blob/main/README.md) has some leaderboards that we could use. I basically kept size 128, and tried both 20 epochs and 200 epochs for each noise level. My reasoning was that it was worth having a somewhat short training run (20 epochs) and a somewhat long training run (200 epochs) for each noise level. I kept size 128 because I assume any observations will transfer over to larger image sizes, and 128x128 images are easier/faster to work with. Of course, if you and Jeremy don't agree, I am willing to change the main leaderboards if necessary.

@jph00 I have already provided a [training example](https://github.com/tmabraham/noisy_imagenette/blob/main/train_noisy_imagenette.py) that works with the CSV and the only thing that would need to be changed is the path once the CSV is added to the tgz and uploaded.

**Comment by KeremTurgutlu** [comment]
Great work @tmabraham! There are usually 2 types of noise benchmarked in the literature, `symmetric noise` and `asymmetric noise`. These functions [here](https://github.com/shengliu66/ELR/blob/108ea3dbcf08d172bec63572ffd32db49c93aca3/ELR/data_loader/cifar10.py#L73) from ELR repo might be helpful for noise generation process for the both types.

What we have so far is `symmetric noise` since we randomly flip labels with **equal_probability** and I believe this is great for start as long as we are transparent about it in the docs. On the other hand, if we would like to add `asymmetric noise` CSVs as well then we could probably do something like:

- Train a model or pick one of the top pretrained models for the desired dataset (ImageWoof or ImageNette).
- Generate confusion matrix from validation set.
- Use the upper triangle, normalize it and use those as probabilities for flipping a label to another.
or
- Take an simpler approach like [here](https://github.com/shengliu66/ELR/blob/108ea3dbcf08d172bec63572ffd32db49c93aca3/ELR/data_loader/cifar10.py#L82) and for each label class flip it to their most confused counterpart.

**Comment by jph00** [comment]
Which is more common in the literature? I assume we should follow whatever is standard.

**Comment by KeremTurgutlu** [comment]
Usually papers report results on both, for example:

```
We show that our framework results in
DNN models with superior generalization performance on CIFAR-10, CIFAR-100 & ImageNet and
outperforms all previous works under symmetric (uniform) and asymmetric noises.
```
from [SELF](https://arxiv.org/pdf/1910.01842.pdf).

Reviewers also care about real world noise;

```
We evaluate the proposed methodology on two standard benchmarks with simulated label noise,
CIFAR-10 and CIFAR-100 [18], and two real-world datasets, Clothing1M [47] and WebVision [24].
For CIFAR-10 and CIFAR-100 we simulate label noise by randomly flipping a certain fraction of
the labels in the training set following a symmetric uniform distribution (as in Eq. (1)), as well as
a more realistic asymmetric class-dependent distribution, following the scheme proposed in [31].
Clothing1M consists of 1 million training images collected from online shopping websites with labels
generated using surrounding text. Its noise level is estimated at 38:5% [36]. For ease of comparison
to previous works [17, 7], we consider the mini WebVision dataset which contains the top 50 classes
from the Google image subset of WebVision, which results in approximately 66 thousand images.
The noise level of WebVision is estimated at 20% [24].
```
from [ELR](https://arxiv.org/pdf/2007.00151.pdf)

As a practitioner I would care whether the benchmarks transfer to the real world or not. Since, it's probably not possible to have a real world noise for woof and nette we can probably go with symmetric and asymmetric for now. Otherwise, we could create mini versions of Clothing1M and WebVision.

**Comment by tmabraham** [comment]
It's up to @jph00 to decide if we should include asymmetric noise. My only concern is that it adds another set of leaderboards and it might be too much. But if we decide to include it, I can make a version of Imagenette with asymmetric noise and add it to my `noisy_imagenette` repository.

**Comment by jph00** [comment]
I'd suggest not including it.

**Comment by KeremTurgutlu** [comment]
Once the leaderboard is official I can work with @tmabraham to test out few papers that we implemented as callbacks and include them as examples/initial results.

**Comment by hamelsmu** [comment]
I'm starting to work on this now, sorry for the delay.  I was a bit lost on some things and @jph00 clarified what I needed to do.  I'm going to first start by familiarizing myself with the leaderboard and the code that uses it so I understand what is going on, and then move forward from there.  

I am not available next week at all, so this may take me a bit.  I'll update my progress on this issue as I make progress

**Comment by tmabraham** [comment]
@hamelsmu Thanks for the update! Let me know if you have any questions.

**Comment by hamelsmu** [comment]
Just for my own background learning, can someone educate me why this is not allowed:

> No inference time tricks, e.g. no: TTA, validation size > train size

Also what is the last bit about the validation size > train size?   Someone would make the validation size greater than the training set size?  Why would they do that?

**Comment by tmabraham** [comment]
@hamelsmu I think this is referring to the _image_ size in the validation set being larger than for the training set. I think there are some papers suggesting there can be an improvement in this case, but I don't remember the details now.

**Comment by KeremTurgutlu** [comment]
I think you are referring to [FixRes](https://arxiv.org/pdf/1906.06423) paper, they argue lower training image size than test improves results. In general, excluding all the test time tricks is to fairly compare different training approaches. Also, TTA has randomness in it which adds extra noise for comparison.

**Comment by hamelsmu** [comment]
@tmabraham do you have a csv file for imagewoof as well?  I only see one for imagenette.  Thanks for your help.

**Comment by tmabraham** [comment]
@hamelsmu Here you go:
https://github.com/tmabraham/noisy_imagenette/blob/main/noisy_imagewoof.csv

**Comment by hamelsmu** [comment]
Sorry I was ignoring your training script, which was my stupid mistake!  

Also I forgot Jeremy said to only add two new leaderboards

10% and 50% imagenette and no imagewoof

**Comment by tmabraham** [comment]
@hamelsmu Okay thanks for letting me know. Do you need me to re-generate the CSV then? I don't have 10% in the CSV. Also, I would have to rerun the baselines. Are we sticking to size 128 and 20/200 epochs, giving a total of 4 leaderboards?

**Comment by hamelsmu** [comment]
Taking some notes in this issue

This will be the directory structure of the new files

```
├── imagenette2
│   ├── noisy_imagenette.csv
│   ├── train
│   └── val
├── imagenette2-160
│   ├── noisy_imagenette.csv
│   ├── train
│   └── val
├── imagenette2-320
│   ├── noisy_imagenette.csv
│   ├── train
│   └── val
├── imagewoof2
│   ├── noisy_imagewoof.csv
│   ├── train
│   └── val
├── imagewoof2-160
│   ├── noisy_imagewoof.csv
│   ├── train
│   └── val
└── imagewoof2-320
    ├── noisy_imagewoof.csv
    ├── train
    └── val
```

**Comment by hamelsmu** [comment]
@tmabraham no no need I'm going with 5% instead of 10% (I talked to Jeremy about this).   Not sure about the size and epochs part but will ask about that thanks for bringing that up

**Comment by hamelsmu** [comment]
@tmabraham I'm thinking I'll just use the same size and epochs as the non-noisy LB.  LMK if you think this sounds like a bad idea for some reason.

![image](https://user-images.githubusercontent.com/1483922/107110923-93916480-6800-11eb-9369-33fbe71712a6.png)

**Comment by jph00** [comment]
@hamelsmu you'll need to update the hashes in fastai too. If you delete the imagenette tgzs from your .fastai/archive directory and your data directory, then call `untar_data` on imagenette, you'll get an error which tells you how to update the hashes. You'll need to do this for each of the versions of the dataset. Let me know if you need a hand. :)

**Comment by hamelsmu** [comment]
> You'll need to do this for each of the versions of the dataset. Let me know if you need a hand. :)

Done in https://github.com/fastai/fastai/pull/3210

**Comment by hamelsmu** [comment]
Ok I believe this is done.  Thanks to @tmabraham for the extensive testing.
